### PR TITLE
rpc,*: sanitize the configuration of rpc.Context

### PR DESCRIPTION
**Pull request by knz**
The main goal of this change is to offer a `.RPCClientConn()` method
on test servers.

To achieve this, it was necessary to lift the code previously in
`cli.getClientGRPCConn` into a more reusable version of it,
now hosted in `rpc.NewClientContext()`.

I also took the opportunity to remove the dependency of `rpc.Context`
on `base.Config`, by spelling out precisely which fields are necessary
to RPC connections.

Numerous tests could be simplified as a result.

Needed for  #107866.
Epic: CRDB-18499

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 7 of 7 files at r5, 58 of 58 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107868)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @herkolategan, @knz, @nkodali, @smg260, and @stevendanna)

___
*[`pkg/base/config.go` line 350 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahGY0G3vj68RVvMFGf:-NahGY0G3vj68RVvMFGg:b-4010o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/base/config.go#L350)):*
> ```Go
> 	// It is set after the server instance has been created, when the
> 	// network listeners are being set up.
> 	AdvertiseAddrH
> ```

For my own edification: what does H stand for? "Host"?
___
*[`pkg/cli/rpc_client.go` line 82 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahF2Q6-wOY8KDlJY0i:-NahF2Q6-wOY8KDlJY0j:b-wd6509) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74238f3c2b49eb081cfeaa99eee34770b1978f2/pkg/cli/rpc_client.go#L82)):*
> ```Go
> }
> 
> func addrWithDefaultHost(addr string) (string, error) {
> ```

nit: the removal of `addrWithDefaultHost` should be in the first commit, not the second.
___
*[`pkg/cli/rpc_client.go` line 28 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahExYH6dsmlnv-CMb0:-NahExYH6dsmlnv-CMb1:bud943a) ([raw file](https://github.com/cockroachdb/cockroach/blob/2546917b685e9ad669d11512314ba4392d0d5bba/pkg/cli/rpc_client.go#L28)):*
> ```Go
> )
> 
> // getClientGRPCConn returns a ClientConn, a Clock and a method that blocks
> ```

nit: the comment on `getClientGRPCConn` needs an update in the first commit.

Update: oh, it's removed in the third commit, up to you.
___
*[`pkg/cli/democluster/demo_cluster.go` line 1077 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahK4nG3dwUdVHGsVf6:-NahK4nG3dwUdVHGsVf7:bi1g37a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/cli/democluster/demo_cluster.go#L1077)):*
> ```Go
> 	// Find a server to use to send the Decommission request.
> 	for _, s := range c.servers {
> 		if s.adminClient != nil && s.nodeID != roachpb.NodeID(nodeID) {
> ```

It seems like previously we would get an admin client to the first node which might be the same as `nodeID`, right? So in a sense we're fixing a bug here - should it be extracted in a separate commit then?
___
*[`pkg/cli/democluster/demo_cluster.go` line 1183 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahKPoj1SxNhTEbqy2b:-NahKPoj1SxNhTEbqy2c:bmv2pxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/cli/democluster/demo_cluster.go#L1183)):*
> ```Go
> 	conn, err := serv.RPCClientConnE(username.RootUserName())
> 	if err != nil {
> 		serv.Stopper().Stop(context.Background())
> ```

nit: we do have `ctx` in scope.
___
*[`pkg/rpc/client.go` line 159 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahM6jmCtRjqvmenFhr:-NahM6jmCtRjqvmenFhs:b-nxz6dw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/rpc/client.go#L159)):*
> ```Go
> 	addr, err := addr.AddrWithDefaultLocalhost(cfg.ServerAddr)
> 	if err != nil {
> 		stopper.Stop(ctx)
> ```

nit: perhaps extract `Stop` into the unconditional deferred function that `Stop`s the stopper in all error cases.
___
*[`pkg/rpc/context.go` line 319 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahMXqD7AsqiWlOK_jO:-NahMXqD7AsqiWlOK_jP:bythypc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/rpc/context.go#L319)):*
> ```Go
> 	// The following fields come from base.Config and only needed by
> 	// SecurityContextOptions. They are not really used by the 'rpc'
> 	// code. They really belong elsehere - see comments in
> ```

nit: `s/elsehere/elsewhere/g`.
___
*[`pkg/server/testserver.go` line 2345 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahPMivDYb1kb0304vS:-NahPMivDYb1kb0304vT:b-vbd7lh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/server/testserver.go#L2345)):*
> ```Go
> ) *rpc.Context {
> 	ctx = logtags.AddTag(ctx, "testclient", nil)
> 	ctx = logtags.AddTag(ctx, "user", user)
> ```

The original code also added nodeID as the tag - should we add SQL instance ID here?
___
*[`pkg/server/testserver.go` line 2362 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahOxp3DdHluYClAMQR:-NahOxp3DdHluYClAMQS:b24z6re) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/server/testserver.go#L2362)):*
> ```Go
> 	ccfg := rpc.MakeClientConnConfigFromBaseConfig(*cfg,
> 		user,
> 		nil, /* tracer */
> ```

Why do we not pass the tracer from `s`? This probably deserves a comment.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107868)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @herkolategan, @nkodali, @smg260, @stevendanna, and @yuzefovich)

___
*[`pkg/base/config.go` line 350 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahGY0G3vj68RVvMFGf:-NahY1elCxYhLNuve8vs:b-pamu64) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/base/config.go#L350)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

For my own edification: what does H stand for? "Host"?
</blockquote></details>

Handle.
I could also call it "container". Unsure what's best.
___
*[`pkg/cli/rpc_client.go` line 82 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahF2Q6-wOY8KDlJY0i:-NahZ1w4E4vBD-WwyHOy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74238f3c2b49eb081cfeaa99eee34770b1978f2/pkg/cli/rpc_client.go#L82)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the removal of `addrWithDefaultHost` should be in the first commit, not the second.
</blockquote></details>

Done.
___
*[`pkg/cli/rpc_client.go` line 28 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahExYH6dsmlnv-CMb0:-NahZ34M0tqYoaVTDJtc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2546917b685e9ad669d11512314ba4392d0d5bba/pkg/cli/rpc_client.go#L28)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the comment on `getClientGRPCConn` needs an update in the first commit.

Update: oh, it's removed in the third commit, up to you.
</blockquote></details>

Done.
___
*[`pkg/cli/democluster/demo_cluster.go` line 1077 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahK4nG3dwUdVHGsVf6:-Nahd3JE8LZpQlB5oxGS:b-4b099a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/cli/democluster/demo_cluster.go#L1077)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It seems like previously we would get an admin client to the first node which might be the same as `nodeID`, right? So in a sense we're fixing a bug here - should it be extracted in a separate commit then?
</blockquote></details>

The firstServer is still special-cased elsewhere so i think the bug remains hidden. But point well taken. I've extracted it in a separate commit.
___
*[`pkg/cli/democluster/demo_cluster.go` line 1183 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahKPoj1SxNhTEbqy2b:-Nahdi1b6IedE0C_zRTK:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/cli/democluster/demo_cluster.go#L1183)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we do have `ctx` in scope.
</blockquote></details>

Fixed.
___
*[`pkg/rpc/client.go` line 159 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahM6jmCtRjqvmenFhr:-NahehQZAYKGrb_yguXY:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/rpc/client.go#L159)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps extract `Stop` into the unconditional deferred function that `Stop`s the stopper in all error cases.
</blockquote></details>

Good idea. Done.
___
*[`pkg/rpc/context.go` line 319 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahMXqD7AsqiWlOK_jO:-Nahem2F3g7C1IuAZJgQ:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/rpc/context.go#L319)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/elsehere/elsewhere/g`.
</blockquote></details>

Fixed.
___
*[`pkg/server/testserver.go` line 2345 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahPMivDYb1kb0304vS:-Nahf5qa5GE_EbxtoAy0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/server/testserver.go#L2345)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The original code also added nodeID as the tag - should we add SQL instance ID here?
</blockquote></details>

Done.
___
*[`pkg/server/testserver.go` line 2362 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahOxp3DdHluYClAMQR:-NahfUm76z3hqOrBAD50:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/server/testserver.go#L2362)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Why do we not pass the tracer from `s`? This probably deserves a comment.
</blockquote></details>

Done.

**Review by stevendanna — COMMENTED** [review]
Overall I like this cleanup.  Don't have much to add over Yahor's existing review.

File: pkg/base/config.go [hunk 1]
```diff
@@ -341,10 +341,13 @@ type Config struct {
...
-	// in the cluster. It should be reachable by all other nodes and should
-	// route to an interface that Addr is listening on.
-	AdvertiseAddr string
+	// AdvertiseAddrH contains the address advertised by the server to
+	// other nodes in the cluster. It should be reachable by all other
+	// nodes and should route to an interface that Addr is listening on.
+	//
+	// It is set after the server instance has been created, when the
+	// network listeners are being set up.
+	AdvertiseAddrH
```
[the remark this answers is not in the document]
Comment (knz) [reply pkg/base/config.go:350]: It's a mixed bag really. Most places take the pointer indeed, but the few that don't seem to be safe.

**Review by yuzefovich — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107868)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @herkolategan, @knz, @nkodali, @smg260, and @stevendanna)

___
*[`pkg/base/config.go` line 350 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NahGY0G3vj68RVvMFGf:-NamXiBa0qYIn4skHoqg:b-3qqdoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b4747de6da0c43254853f2f1eb5be6f3d9d68fe/pkg/base/config.go#L350)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Handle.
I could also call it "container". Unsure what's best.
</blockquote></details>

No opinion from me.
___
*[`pkg/cli/democluster/demo_cluster.go` line 1028 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NamX46S2OtHm6fUZCZ9:-NamX46S2OtHm6fUZCZA:b-wqtd4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/1314452b1058dac8bfd1aac3e411499ed6d34bfa/pkg/cli/democluster/demo_cluster.go#L1028)):*
> ```Go
> }
> 
> // findOtherServer returns a an admin RPC client to another
> ```

nit: "a an admin".
___
*[`pkg/rpc/client.go` line 163 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NamZGvP4CNvARUenu9O:-NamZGvP4CNvARUenu9P:bi0hoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04f92fc62c0e19e4056d0b29d205c269add35f8/pkg/rpc/client.go#L163)):*
> ```Go
> 		// even if the context used to create the client conn is
> 		// canceled.
> 		stopper.Stop(context.Background())
> ```

Switching to `context.Background()` seems like a non-trivial change to me (granted, I'm not too familiar with this code). I did some archeology to see where usage of `ctx` and the comment about "tying the lifetime" was introduced and landed on these two ([1](7a3cdaa01834915a05dbd027e7d131dddcca8ea5), [2](da5dca5a528245a20f4255105cbb34c816876ba7)) commits. Can you provide some more context for why we're making `ctx -> context.Background()` switch? Perhaps someone from replication team (perhaps @erikgrinaker) could take a quick look at it too?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107868)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, @herkolategan, @nkodali, @smg260, @stevendanna, and @yuzefovich)

___
*[`pkg/cli/democluster/demo_cluster.go` line 1028 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NamX46S2OtHm6fUZCZ9:-NamfsFX4D6WGYI5G3HF:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1314452b1058dac8bfd1aac3e411499ed6d34bfa/pkg/cli/democluster/demo_cluster.go#L1028)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: "a an admin".
</blockquote></details>

Fixed.
___
*[`pkg/rpc/client.go` line 163 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NamZGvP4CNvARUenu9O:-NamgvXH9KJqSd74raal:b-t10bgu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04f92fc62c0e19e4056d0b29d205c269add35f8/pkg/rpc/client.go#L163)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Switching to `context.Background()` seems like a non-trivial change to me (granted, I'm not too familiar with this code). I did some archeology to see where usage of `ctx` and the comment about "tying the lifetime" was introduced and landed on these two ([1](7a3cdaa01834915a05dbd027e7d131dddcca8ea5), [2](da5dca5a528245a20f4255105cbb34c816876ba7)) commits. Can you provide some more context for why we're making `ctx -> context.Background()` switch? Perhaps someone from replication team (perhaps @erikgrinaker) could take a quick look at it too?
</blockquote></details>

This latter commit da5dca5a528245a20f4255105cbb34c816876ba7 from Tobias introduced a bug. Consider:
```diff
@@ -939,31 +941,31 @@ func getClientGRPCConn(ctx context.Context) (*grpc.ClientConn, *hlc.Clock, error
        stopper.AddCloser(stop.CloserFn(func() {
                _ = conn.Close()
        }))

        // Tie the lifetime of the stopper to that of the context.
-       go func() {
-               <-ctx.Done()
-               // Don't use the `ctx` as it's already cancelled.
-               stopper.Stop(context.Background())
-       }()
-       return conn, clock, nil
+       closer := func() {
+               stopper.Stop(ctx)
+       }
+       return conn, clock, closer, nil
 }
```

Observe:  the original code (introduced in 7a3cdaa01834915a05dbd027e7d131dddcca8ea5) was doing two separate things:
1) waiting for the context to be cancelled before calling .Stop"
2) using a separate context for the call to `.Stop` to ensure the already-cancelled context does not get passed to the closers.

Point (1) was considered to be a problem - instead we want the caller to call `finish()` explicitly.
While that is true (and achieved by Tobi's patch), it introduced a _new_ problem: the `ctx` is passed directly to `.Stop` now. This defeats the protection of point (2).

My PR here is restoring that protection.

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-:-Namq2AE3nj54MjLQzBZ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, @herkolategan, @nkodali, @smg260, and @stevendanna)

___
*[`pkg/rpc/client.go` line 163 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NamZGvP4CNvARUenu9O:-NampXh3B8spb6xUFZ3x:b-wjgsqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04f92fc62c0e19e4056d0b29d205c269add35f8/pkg/rpc/client.go#L163)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

This latter commit da5dca5a528245a20f4255105cbb34c816876ba7 from Tobias introduced a bug. Consider:
```diff
@@ -939,31 +941,31 @@ func getClientGRPCConn(ctx context.Context) (*grpc.ClientConn, *hlc.Clock, error
        stopper.AddCloser(stop.CloserFn(func() {
                _ = conn.Close()
        }))

        // Tie the lifetime of the stopper to that of the context.
-       go func() {
-               <-ctx.Done()
-               // Don't use the `ctx` as it's already cancelled.
-               stopper.Stop(context.Background())
-       }()
-       return conn, clock, nil
+       closer := func() {
+               stopper.Stop(ctx)
+       }
+       return conn, clock, closer, nil
 }
```

Observe:  the original code (introduced in 7a3cdaa01834915a05dbd027e7d131dddcca8ea5) was doing two separate things:
1) waiting for the context to be cancelled before calling .Stop"
2) using a separate context for the call to `.Stop` to ensure the already-cancelled context does not get passed to the closers.

Point (1) was considered to be a problem - instead we want the caller to call `finish()` explicitly.
While that is true (and achieved by Tobi's patch), it introduced a _new_ problem: the `ctx` is passed directly to `.Stop` now. This defeats the protection of point (2).

My PR here is restoring that protection.
</blockquote></details>

Thanks, makes sense, I can sign off on this now. Do you think it's worth extracting this into a separate commit that is also backported? Perhaps we'll fix some extremely rare flakes? Perhaps it's worth extracting it into a separate commit regardless of the backport in order to provide the explanation from your message? Anyway, up to you.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107868)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, @herkolategan, @nkodali, @smg260, @stevendanna, and @yuzefovich)

___
*[`pkg/rpc/client.go` line 163 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/107868#-NamZGvP4CNvARUenu9O:-Nan2wnuEd-q4rY3IAKd:b-w5qp5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04f92fc62c0e19e4056d0b29d205c269add35f8/pkg/rpc/client.go#L163)):*
> it's worth extracting this into a separate commit

ok, sure. Done.

> that is also backported?

I'm not so sure about that. Until 23.2 (not released yet) all the contexts used in cli code were not cancellable, so this didn't hit.

**Comment by knz** [comment]
TFYRs!

bors r=yuzefovich,stevendanna

**Comment by knz** [comment]
bors single on
bors r=yuzefovich,stevendanna

**Comment by knz** [comment]
bors r=yuzefovich,stevendanna

**Comment by knz** [comment]
bors r=yuzefovich,stevendanna

**Pull request closed without merging**